src/main/java/org/opensearch/knn/index/query/filtered/FilteredKNNIterator.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.filtered;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.opensearch.knn.index.SpaceType;
+
+import java.io.IOException;
+
+/**
+ * Inspired by DiversifyingChildrenFloatKnnVectorQuery in lucene
+ * https://github.com/apache/lucene/blob/7b8aece125aabff2823626d5b939abf4747f63a7/lucene/join/src/java/org/apache/lucene/search/join/DiversifyingChildrenFloatKnnVectorQuery.java#L162
+ *
+ * The class is used in KNNWeight to score filtered KNN field by iterating filterIdsArray.
+ */
+public abstract class FilteredKNNIterator<T extends KNNQueryVector> {
+    // Array of doc ids to iterate
+    protected final int[] filterIdsArray;
+    protected float currentScore = Float.NEGATIVE_INFINITY;
+    protected final T queryVector;
+    protected final BinaryDocValues values;
+    protected final SpaceType spaceType;
+    protected int currentPos = 0;
```

Comment:
Any reason not use getter and setter method to access from sub class instead of exposing variable?

Reply:
Those variables are heavily accessed in child class. Two reasons. 1. More efficient because there is not method call involved. 2. `filterIdsArray[currentPos]` and `currentPos++` are more readable than `getFilterIdsArray()[getCurrentPos())]` and `setCurrentPos(getCurrentPos() + 1)`

Reply:
IMO i don't think it really matters in terms of performance whether you are using protected variable or protected method. I think protected method is better for following reasons: 
1. If you see variable that is used but it is not declared you have to hunt for the variable. 
2. Protected method provides encapsulation for variables. 
3. It does not violate Open Closed principle.

If you have operations like currentPoss++, setCurrentPos(get + 1) and you think it is used heavily, i would create a methods in abstract class with proper name to avoid chaining

Reply:
1. Same for method. IDE is there for hunting the variable.
2. I don't need encapsulation of the variable

Anyway, I updated to use the getter, setter. I feel like it somehow reduced the readability. It is something like you are accessing your own variable through getter/setter method. 
For example,
```
@Getter(AccessLevel.Protected)
public class MyClass {
  private int myVariable;

  public int getNewValue() {
    return getMyVariable() + 1;
  }
}
```

Reply:
@VijayanB I would agree with @heemin32 here. Getters and setters encapsulations are good if the variables are actually used outside of the class. Having the child class use the getters for variables feels like a private attribute of a class getting accessed in the same class via a getter/setter. 

If attributes would have been complex classes which requires init/some checks before accessing it would have made more sense to use getters/setters. 

My recommendation here would be to go back to direct access of the variables.